### PR TITLE
MAINT: spatial.distance.jaccard: correct numerator calculation

### LINKED ISSUE #14304: Jaccard distance greater than 1 if elements are strings
```
>python
Python 3.9.5 (default, May 18 2021, 14:42:02) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import scipy 
>>> scipy.__version__
'1.6.2'
>>> from scipy.spatial.distance import jaccard
>>> jaccard([1, 0, 0], [0, 1, 0])
1.0
>>> jaccard(["A", "B", "B"], ["B", "A", "B"])
FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  nonzero = np.bitwise_or(u != 0, v != 0)
2.
```

### PR DESCRIPTION
**Pull request by fancidev**
#### Reference issue
Follows up gh-14357. Closes gh-14304, gh-21158.

#### What does this implement/fix?
The Jaccard distance between boolean vectors $(u_1, \cdots, u_n)$ and $(v_1, \cdots, v_n)$ is defined as 

$$
\frac{\sum u_i \textrm{XOR} v_i}{\sum u_i \textrm{OR} v_i }
$$

For non-Boolean and non-0/1 input, the current code computes the denominator by casting the input to bool, but it computes the numerator in an ad-hoc way that is not consistent with the definition of Jaccard distance. This PR corrects this by casting the input to bool before computing the numerator.

I also take the chance to enrich the doc of Jaccard, to clarify the definitions and address some of the confusion from past issues (mostly with regard to input being original observations or encoded observations). Other editorial changes include:

- removed Needham from the doc as I wasn't able to find any source citing the metric is related to Needham
- the doc had mixed use of "1-D arrays" and "vectors"; I change all to "vectors" for consistency
- change $c_{TF}$ to $c_{10}$ etc as it works easier with the formula and is aligned with *some* of the other metrics
- the doc had mixed use of "Jaccard distance" and "Jaccard dissimilarity"; I change all to "Jaccard dissimilarity" for consistency
- add a paragraph with weighted formula
- remove an example with non-0/1 input as I consider it not advertised usage
- add a book to the reference section -- its first chapter contains a definition of all of the Boolean metrics in spatial, and with much context and discussions
- add an example of weighted dissimilarity

#### Additional information
I leave the behavior of non-Boolean, non-0/1 input unspecified in the doc. The reason is that there remain a few inconsistencies between metrics with respect to input handling. My goal is to align them eventually (in two years' time).

### CODE REVIEW

**Comment by lucascolley** [comment]
can we add a unit test :) ?

**Comment by fancidev** [comment]
> can we add a unit test :) ?

Added!

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -770,75 +770,118 @@ def hamming(u, v, w=None):
...
+    For example, subsets :math:`A,B` of :math:`\{ 1, 2, ..., n \}` may be
+    encoded into boolean vectors :math:`u, v` by setting
+    :math:`u_k := 1_{k \in A}`, :math:`v_k := 1_{k \in B}`
+    for :math:`k = 1,2,\cdots,n`.
 
     .. versionchanged:: 1.2.0
-        Previously, when `u` and `v` lead to a ``0/0`` division, the function
-        would return NaN. This was changed to return 0 instead.
+        Previously, when `u` and `v` are both zero vectors, the function
+        would return ``nan``. This was changed to return ``0`` instead.
```
Comment (lucascolley) [inline scipy/spatial/distance.py:836]: is it true that we can only get `0/0` if `u` and `v` are both zero vectors? What about if some of the weights are `0`?

File: scipy/spatial/src/distance_metrics.h [hunk 2]
```diff
@@ -449,7 +449,7 @@ struct JaccardDistance {
     void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
         transform_reduce_2d_<4>(out, x, y, [](T x, T y) INLINE_LAMBDA {
             Acc<T> acc;
-            acc.num = (x != y) & ((x != 0) | (y != 0));
+            acc.num = (x != 0) ^ (y != 0);
```
Comment (lucascolley) [inline scipy/spatial/src/distance_metrics.h:452]: the changes in this file have no impact on the Python function, right? If so, could you put them in a separate commit?

File: scipy/spatial/tests/test_distance.py [hunk 3]
```diff
@@ -2273,3 +2205,84 @@ def test_immutable_input(metric):
...
+        Y_right = eo['pdist-jaccard']
+        Y_test1 = wpdist(X, 'jaccard')
+        assert_allclose(Y_test1, Y_right, rtol=eps)
+
+    def test_pdist_djaccard_random_float32(self):
+        eps = 1e-8
+        X = np.float32(eo['pdist-boolean-inp'])
+        Y_right = eo['pdist-jaccard']
+        Y_test1 = wpdist(X, 'jaccard')
+        assert_allclose(Y_test1, Y_right, rtol=eps)
```
Comment (lucascolley) [inline scipy/spatial/tests/test_distance.py:2245]: not really related to this PR, but there is an opportunity to parametrize over the dtype in these tests

**Review by lucascolley — COMMENTED** [review]
sorry @fancidev , this fell off my radar, but it looks great! It seems to line up nicely with what @rkern was after in the linked PR. Just a few comments.

@jakevdp , if you would like to take a quick look at the doc changes here, that would be much appreciated!

File: scipy/spatial/distance.py (hunk 1 above)
Comment (fancidev) [reply scipy/spatial/distance.py:836]: Indeed what I put above is only accurate for the unweighted case. For the weighted case, the function would return `nan` if all positive-weighted elements in u and v are zero. Let me update it in the doc.

File: scipy/spatial/src/distance_metrics.h (hunk 2 above)
Comment (fancidev) [reply scipy/spatial/src/distance_metrics.h:452]: Correct, the C++ change and the Python change are independent.

Let me try rearrange the commits to "one at a time".

File: scipy/spatial/tests/test_distance.py (hunk 3 above)
Comment (fancidev) [reply scipy/spatial/tests/test_distance.py:2245]: I too think there's a lot room for improvement in the existing tests. I only did a small refactoring (by putting all tests related to Jaccard distance together). A systematic review of the tests (coverage, set-up, de-duplication, etc) could be a good topic for a separate PR!

**Comment by fancidev** [comment]
Thanks for reviewing @lucascolley ! I rearranged the commits (to one change at a time), and updated the "version-changed" message in the doc per your suggestion. Hope it looks better now!

**Review by lucascolley — APPROVED** [review]
Thanks @fancidev ! I will leave this open for a while in case there are any other comments about the docs. Otherwise, will merge before next release.

**Review by lucascolley — COMMENTED** [review]
> I leave the behavior of non-Boolean, non-0/1 input unspecified in the doc.

Actually, I wonder whether we should still add a `version changed` admonition to explain the changed behaviour for non-0/1 input. On the one hand, anyone using the function that way was clearly misusing it. But it may be helpful for debugging regardless. I think that would be okay as long as we keep the parameters as officially taking array-likes of bools.

**Comment by fancidev** [comment]
> Actually, I wonder whether we should still add a `version changed` admonition to explain the changed behaviour for non-0/1 input. On the one hand, anyone using the function that way was clearly misusing it. But it may be helpful for debugging regardless. I think that would be okay as long as we keep the parameters as officially taking array-likes of bools.

Good idea! How about adding a “version changed” message like the following:

Changed in version 1.15.0: Non-0/1 numeric input used to produce an ad hoc result. Since 1.15.0, numeric input is converted to Boolean before computation.

**Comment by lucascolley** [comment]
> How about adding a “version changed” message like the following:
>
> Changed in version 1.15.0: Non-0/1 numeric input used to produce an ad hoc result. Since 1.15.0, numeric input is converted to Boolean before computation.

sounds good to me!

**Review by lucascolley — APPROVED** [review]

**Comment by fancidev** [comment]
I resolved a merge conflict and fix a lint warning and a smoke doc warning. 

@lucascolley Do you think the current state is ok for merge?

**Comment by lucascolley** [comment]
thanks again!
